File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <bits/stdc++.h>
using namespace std;
typedef vector<vector<int>> vi;
const int N = 20002;
vector<int> g[N];
int dist[N];

int solution(int n, vi edge) 
{
    memset(dist, -1, sizeof dist);
    for (auto to : edge)
    {
        int u = to[0], v = to[1];
        g[u].push_back(v);
        g[v].push_back(u);
    }

    dist[1] = 0; queue<int> q; q.push(1);
    while (!q.empty())
    {
        int v = q.front(); q.pop();
        for (int nv : g[v])
        {
            if (dist[nv] != -1) continue;
            q.push(nv);
            dist[nv] = dist[v] + 1;
        }
    }
    
    int mx = 0;
    
    for (int i = 1; i <= n; i++)
        mx = max(mx, dist[i]);
    
    int ans = 0;

    for (int i = 1; i <= n; i++)
    {
        if (dist[i] == mx)
            ans++;
    }
    
    return ans;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution 
{
    static final int N = 20002;
    ArrayList<Integer>[] g = new ArrayList[N];
    int[] dist = new int[N];
    
    public int solution(int n, int[][] edge) 
    {
        for (int i = 1; i <= n; i++)
        {
            dist[i] = -1;
            g[i] = new ArrayList<>();
        }
        
        int ans = 0;
        int len = edge.length;
        for (int i = 0; i < len; i++)
        {
            int u = edge[i][0], v = edge[i][1];
            g[u].add(v); g[v].add(u);
        }
        
        Queue<Integer> q = new LinkedList<>(); 
        q.add(1); dist[1] = 0;
        while (!q.isEmpty())
        {
            int v = q.peek(); q.poll();
            for (int nv : g[v])
            {
                if (dist[nv] == -1)
                {
                    dist[nv] = dist[v] + 1;
                    q.add(nv);
                }
            }
        }
        
        int mx = -1, cnt = 0;
        for (int v = 1; v <= n; v++)
        {
            if (mx < dist[v])
            {
                mx = dist[v];
                cnt = 1;
            }
            else if (mx == dist[v])
                cnt++;
        }
        
        return cnt;
    }
}